sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2048,19 +2088,34 @@ class Dataset[T] private[sql](
     // Builds a project list for `other` based on `logicalPlan` output names
     val rightProjectList = leftOutputAttrs.map { lattr =>
       rightOutputAttrs.find { rattr => resolver(lattr.name, rattr.name) }.getOrElse {
-        throw new AnalysisException(
-          s"""Cannot resolve column name "${lattr.name}" among """ +
-            s"""(${rightOutputAttrs.map(_.name).mkString(", ")})""")
+        if (allowMissingColumns) {
```

Comment:
Does it work with nested columns?

Reply:
No, currently it doesn't.

Reply:
I think the major problem here is we put the by-name logic in the API method, not in the `Analyzer`. Shall we add a boolean parameter to `Union`, and move the by-name logic to the type coercion rules?

Reply:
Ok. I will do it in another PR.

Reply:
@cloud-fan . `unionByName` has been here since Apache Spark 2.3.0. 
It would be great if we can proceed that refactoring suggestion as a separate JIRA.

Reply:
Yea it's better to have a new JIRA.

Reply:
Thanks, @cloud-fan .